### ISSUE TITLE
Custom export takes too long

### ISSUE THREAD

**Issue by patrikvalentaa** [contributte/datagrid#1146]
Hi @f3l1x, @paveljanda,

Is possible to change method getDataSource() in file Contributte\Datagrid\DataSource\NextrasDataSource.php [here](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/DataSource/NextrasDataSource.php#L229) visibility from protected to public?

In my case I am doing custom export, because included CSV export takes too long in case large model dependences, so I prepared custom SQL query and now i need to pass argument $userIds, which I get from filtered items from datagrid. For now there is no accessable method to get collection, here is only method getData(), but it returns Nextras Model fetchAll(), that i later iterating for get array of filtered ids. Which also takes too long. Best way for me is public visibility method getDataSource(), where I can call fetchPairs(null, 'id').

Thanx

**Comment by sasule** [comment]
Seems to me like you are exporting all data and then want to filter them by filtered rows from grid.

Wouldn't be method ` $grid->addExportCsvFiltered()` a solution? 

See [docs](https://contributte.org/packages/contributte/datagrid/export.html#exportcallback).

**Comment by patrikvalentaa** [comment]
Not exactly. I used to use $grid->addExportCsvFiltered(), but there is same problem with fetching all amount of data and model dependecies such as relationships, virtual properities, etc. 

If I use [addExportCsvFiltered](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/Datagrid.php#L1557), it creates new object [ExportCsv](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/Datagrid.php#L1577), which contains public method [isFiltered()](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/Export/Export.php#L138), that method is called in [DataGrid.php](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/Datagrid.php#L1837) at method handleExport. That assembly the filters and set them in [DataModel.php](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/Datagrid.php#L1851), which returns method [getData()](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/DataModel.php#L130C5-L130C40) or [here](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/DataModel.php#L133). Here comes the problem with calling getData() method, because it fire [fetchAll](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/DataSource/NextrasDataSource.php#L46) method, but this is the problematic part which make my export slow. The solution is make accessible method [getDataSource()](https://github.com/contributte/datagrid/blob/0105ca358f9c945874ae464b36b131d1915a5644/src/DataSource/NextrasDataSource.php#L229), because DataGrid makes the whole ICollection updated, based on ordering or filtering via DateTimeFilters or TextFilters or any other supported types of filters. So, if I getData(), which contains array of Entity I need to make iteration, where i collect ids to other property. But if can access getDataSource of the current Datagrid datasource (in my case NextrasDataSource), it returns in my case the ICollection object. Then I can call fetchPairs(null, 'id'), which can get me array of integers (Entity IDs), later I pass them to my SQL query, which returns me exact items as I can get from addExportCsvFiltered(), but with one important difference. Now export won't take 3 minutes, but 3 seconds. I tried this locally by modifying it in vendor. I have large database with amount of dependencies on other Entities, which are slower down whole ORM and whole process.

**Comment by radimvaculik** [comment]
`getDataSource` is already public.

See https://github.com/contributte/datagrid/blob/master/src/DataSource/NextrasDataSource.php#L113 👀
